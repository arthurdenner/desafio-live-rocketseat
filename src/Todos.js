import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { isEmpty } from "lodash/fp";

const Todos = ({ todos }) =>
  isEmpty(todos) ? (
    <div className="emptyTodos">There is no itens in the list</div>
  ) : (
    <ul>{todos.map(todo => <li key={todo.id}>{todo.text}</li>)}</ul>
  );

Todos.propTypes = {
  todos: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  todos: state.todos
});

export default connect(mapStateToProps)(Todos);
