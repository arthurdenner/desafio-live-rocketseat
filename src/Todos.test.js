import React from "react";
import { shallow } from "enzyme";
import configureStore from "redux-mock-store";
import Todos from "./Todos";

// configure mock store
const initialState = {
  todos: [
    { id: 1, text: "Watch the live " },
    { id: 2, text: "Do the challenge " }
  ]
};
const mockStore = configureStore();
const store = mockStore(initialState);

// since we are changing props on some tests,
// it's better to initialize the wrapper as a let
// and reinitialize it before each test
let wrapper;

describe("todos component", () => {
  beforeEach(() => {
    wrapper = shallow(<Todos />, { context: { store } }).dive();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render only one ul element", () => {
    expect(wrapper.find("ul").length).toEqual(1);
  });

  it("should render at least one li element if the todos prop is NOT empty", () => {
    expect(wrapper.find("li").length).toBeGreaterThanOrEqual(1);
  });

  it("should render a div with a className emptyTodos if the todos prop is empty", () => {
    // it's necessary to use setProps to realize the test
    // otherwise, it would consider the prop provided by the connect function
    wrapper.setProps({ todos: [] });

    expect(wrapper.find(".emptyTodos").length).toEqual(1);
  });

  it("should render a text saying that there is no itens if the todos prop is empty", () => {
    // it's necessary to use setProps to realize the test
    // otherwise, it would consider the prop provided by the connect function
    wrapper.setProps({ todos: [] });

    const emptyTodosDiv = wrapper.find(".emptyTodos").first();

    expect(emptyTodosDiv.text()).toEqual("There is no itens in the list");
  });
});
