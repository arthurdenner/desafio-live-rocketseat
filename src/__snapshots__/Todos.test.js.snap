// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todos
    dispatch={[Function]}
    todos={
        Array [
            Object {
              "id": 1,
              "text": "Watch the live ",
            },
            Object {
              "id": 2,
              "text": "Do the challenge ",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li>
          Watch the live 
</li>,
        <li>
          Do the challenge 
</li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": "Watch the live ",
        },
        "ref": null,
        "rendered": "Watch the live ",
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": "Do the challenge ",
        },
        "ref": null,
        "rendered": "Do the challenge ",
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li>
            Watch the live 
</li>,
          <li>
            Do the challenge 
</li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": "Watch the live ",
          },
          "ref": null,
          "rendered": "Watch the live ",
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": "Do the challenge ",
          },
          "ref": null,
          "rendered": "Do the challenge ",
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
